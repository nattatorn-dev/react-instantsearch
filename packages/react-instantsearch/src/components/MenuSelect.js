import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { find } from 'lodash';

import classNames from './classNames.js';
import translatable from '../core/translatable';

const cx = classNames('MenuSelect');

class MenuSelect extends Component {
  static propTypes = {
    canRefine: PropTypes.bool.isRequired,
    refine: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired,
        count: PropTypes.number.isRequired,
        isRefined: PropTypes.bool.isRequired,
      })
    ),
  };

  static contextTypes = {
    canRefine: PropTypes.func,
  };

  get selectedValue() {
    const { value } = find(this.props.items, { isRefined: true }) || {
      value: 'ais__see__all__option',
    };
    return value;
  }

  componentWillMount() {
    if (this.context.canRefine) this.context.canRefine(this.props.canRefine);
  }

  componentWillReceiveProps(props) {
    if (this.context.canRefine) this.context.canRefine(props.canRefine);
  }

  handleSelectChange = ({ target: { value } }) => {
    this.props.refine(value === 'ais__see__all__option' ? '' : value);
  };

  render() {
    const { items, translate } = this.props;

    return (
      <select
        value={this.selectedValue}
        onChange={this.handleSelectChange}
        {...cx('select')}
      >
        <option value="ais__see__all__option" {...cx('option')}>
          {translate('seeAllOption')}
        </option>

        {items.map(item => (
          <option key={item.value} value={item.value} {...cx('option')}>
            {item.label} ({item.count})
          </option>
        ))}
      </select>
    );
  }
}

export default translatable({
  seeAllOption: 'See all',
})(MenuSelect);
