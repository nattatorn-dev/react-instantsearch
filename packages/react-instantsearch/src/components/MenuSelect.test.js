/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import MenuSelect from './MenuSelect';

describe('MenuSelect', () => {
  it('default menu select', () => {
    const tree = renderer
      .create(
        <MenuSelect
          refine={() => {}}
          items={[
            { label: 'white', value: 'white', count: 10, isRefined: false },
            { label: 'black', value: 'black', count: 20, isRefined: false },
            { label: 'blue', value: 'blue', count: 30, isRefined: false },
            { label: 'green', value: 'green', count: 30, isRefined: false },
            { label: 'red', value: 'red', count: 30, isRefined: false },
          ]}
          canRefine={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies translations', () => {
    const tree = renderer
      .create(
        <MenuSelect
          refine={() => {}}
          items={[
            { label: 'white', value: 'white', count: 10, isRefined: false },
            { label: 'black', value: 'black', count: 20, isRefined: false },
            { label: 'blue', value: 'blue', count: 30, isRefined: false },
            { label: 'green', value: 'green', count: 30, isRefined: false },
            { label: 'red', value: 'red', count: 30, isRefined: false },
          ]}
          translations={{
            seeAllOption: 'Everything',
          }}
          canRefine={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <MenuSelect
        refine={refine}
        items={[
          { label: 'white', value: 'white', count: 10, isRefined: false },
          { label: 'black', value: 'black', count: 20, isRefined: false },
          { label: 'blue', value: 'blue', count: 30, isRefined: false },
        ]}
        canRefine={true}
      />
    );

    const items = wrapper.find('.ais-MenuSelect__option');
    expect(items.length).toBe(4); // +1 from "see all option"

    wrapper
      .find('.ais-MenuSelect__select')
      .simulate('change', { target: { value: 'blue' } });

    expect(refine).toHaveBeenCalledTimes(1);
    expect(refine).toHaveBeenCalledWith('blue');

    wrapper.unmount();
  });
});
