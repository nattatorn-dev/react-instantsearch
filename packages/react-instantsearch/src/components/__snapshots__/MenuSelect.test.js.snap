// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuSelect applies translations 1`] = `
<select
  className="ais-MenuSelect__select"
  onChange={[Function]}
  value="ais__see__all__option"
>
  <option
    className="ais-MenuSelect__option"
    value="ais__see__all__option"
  >
    Everything
  </option>
  <option
    className="ais-MenuSelect__option"
    value="white"
  >
    white
     (
    10
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="black"
  >
    black
     (
    20
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="blue"
  >
    blue
     (
    30
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="green"
  >
    green
     (
    30
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="red"
  >
    red
     (
    30
    )
  </option>
</select>
`;

exports[`MenuSelect default menu select 1`] = `
<select
  className="ais-MenuSelect__select"
  onChange={[Function]}
  value="ais__see__all__option"
>
  <option
    className="ais-MenuSelect__option"
    value="ais__see__all__option"
  >
    See all
  </option>
  <option
    className="ais-MenuSelect__option"
    value="white"
  >
    white
     (
    10
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="black"
  >
    black
     (
    20
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="blue"
  >
    blue
     (
    30
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="green"
  >
    green
     (
    30
    )
  </option>
  <option
    className="ais-MenuSelect__option"
    value="red"
  >
    red
     (
    30
    )
  </option>
</select>
`;
